compiler/src/iree/compiler/Codegen/Common/GPU/GPUInferMemorySpace.cpp

```diff
@@ -0,0 +1,127 @@
+#include "llvm/ADT/STLExtras.h"
+#include "mlir/Dialect/Bufferization/IR/Bufferization.h"
+#include "mlir/Dialect/GPU/IR/GPUDialect.h"
+#include "mlir/IR/Matchers.h"
+#include "mlir/IR/Visitors.h"
+#include "mlir/Interfaces/FunctionInterfaces.h"
+
+namespace mlir::iree_compiler {
+
+#define GEN_PASS_DEF_GPUINFERMEMORYSPACEPASS
+#include "iree/compiler/Codegen/Common/GPU/Passes.h.inc"
+
+namespace {
+
+/// Pass to infer the memory spaces of unmarked `bufferization.alloc_tensor`
+/// ops. Inferring the memory space during bufferization (in the allocation
+/// function) is infeasible due to some limited analysis of surrounding loop
+/// structures needed. After this pass, any unexpected allocations are then
+/// treated as a compiler failure indicating something went wrong during or
+/// near bufferization.
+struct GPUInferMemorySpacePass final
+    : impl::GPUInferMemorySpacePassBase<GPUInferMemorySpacePass> {
+
+  void runOnOperation() override;
+};
+
+bool isDefinitelyThreadLocal(bufferization::AllocTensorOp alloc) {
+  ArrayRef<int64_t> allocShape = alloc.getType().getShape();
+  // Give up on dynamic shapes because we can't easily verify that
+  // the destination is overwritten.
+  if (ShapedType::isDynamicShape(allocShape)) {
+    return false;
+  }
+
+  // An allocation can safely be allocated as private if from within a
+  // distributed context all threads overwrite the whole allocation. The logic
+  // below checks for a limited version of this by only looking for
+  // `vector.transfer_write` ops that fully overwrite the tensor.
+  for (auto user : alloc->getUsers()) {
+    if (!operationHasParentForallOfMappingType<gpu::GPUThreadMappingAttr,
+                                               IREE::GPU::LaneIdAttr>(user)) {
+      return false;
+    }
+    auto write = dyn_cast<vector::TransferWriteOp>(user);
+    // TODO: look through reshapes and linalg op destinations if necessary.
+    if (!write) {
+      return false;
+    }
+
+    ArrayRef<int64_t> sourceVecShape = write.getVectorType().getShape();
+    if (!llvm::all_of_zip(allocShape, sourceVecShape,
+                          [](int64_t l, int64_t r) { return l == r; })) {
+      return false;
+    }
+
+    if (!llvm::all_of(write.getIndices(), [](Value value) {
+          return getConstantIntValue(value) == static_cast<int64_t>(0);
+        })) {
+      return false;
+    }
+  }
+  return true;
+}
+
+bool isDefinitelyShared(bufferization::AllocTensorOp alloc) {
+  // An allocation can be inferred as shared if it is the destination of a
+  // thread distributed `scf.forall` op. All other shared allocations are
+  // expected to be properly indicated in advance.
+  for (auto user : alloc->getUsers()) {
+    auto forallOp = dyn_cast<scf::ForallOp>(user);
+    if (!forallOp ||
+        !forallOpHasMappingType<gpu::GPUThreadMappingAttr,
+                                gpu::GPUWarpMappingAttr>(forallOp)) {
+      return false;
+    }
+  }
+  return true;
+}
```

Comment:
Should this check for cases where there is an alloc used in a forall but isn't the iter_arg? I'm not sure if we ever generate that (aside from TileAndFuse which already will have a memory space).

Reply:
A user of an operation means that it is an explicit operand of that op. The only tensor operand of an `scf.forall` is the `shared_outs`, so just checking the kind of user is good enough.